Adding fix and tests for spherical mesh as spatial distribution

pshriwise opened the pull request:
<!--
If you are a first-time contributor to OpenMC, please have a look at our
contributing guidelines:
https://github.com/openmc-dev/openmc/blob/develop/CONTRIBUTING.md
-->

# Description

This passes the cosine of theta bound for a spherical mesh bin to `uniform_distribution` instead of the angles themselves. It also adds tests using regular, cylindrical, and spherical meshes with `MeshSpatial` with some checks to ensure we're getting valid source values. To simplify said tests I've also added a `SphericalMesh.from_domain` classmethod.

Small additional change: the `origin` attribute is now written as part of the `SphericalMesh.__repr__`.

Fixes #3427

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] ~I have made corresponding changes to the documentation (if applicable)~
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
<!--
While tests will automatically be checked by CI, it is good practice to
ensure that they pass locally first. See instructions here:
https://docs.openmc.org/en/latest/devguide/tests.html
-->

gonuke commented on openmc/mesh.py:
```diff
@@ -2167,6 +2171,84 @@ def from_hdf5(cls, group: h5py.Group, mesh_id: int, name: str):
+            outer_radius,
+            num=dimension[0]+1
+        )
+        theta_grid = np.linspace(
+            theta_grid_bounds[0],
+            theta_grid_bounds[1],
+            num=dimension[1]+1
+        )
```
Wouldn't you want the default behavior for these bounds to be uniform in cos(theta)?

pshriwise replied:
Right. The cosine of theta could be better for certain applications, but I think I'd leave that to the user to replace the grid in those situations really. Same could be said about applying a power law to the the `r_grid` spacing if we're looking to make volumes equal or something.

Let me know if you feel strongly otherwise.

pshriwise commented on openmc/mesh.py:
```diff
@@ -2167,6 +2171,84 @@ def from_hdf5(cls, group: h5py.Group, mesh_id: int, name: str):
+        # loaded once to avoid recalculating bounding box
+        cached_bb = domain.bounding_box
+
+        outer_radius = 0.5 * max(cached_bb.width)
```
So, again, a matter perhaps of expectation. Our `CylindricalMesh.from_domain` method ensures that the domain is encompassed by the mesh, but I can definitely see cases (especially for spherical meshes) where the intent would be to have the mesh inscribe the bounding box of the provided domain. An option for this could easily be provided for each and it wouldn't take much additional testing. Thoughts?

pshriwise commented:
Like a vine on the wall, the scope has crept after thinking over some of the discussion above.

I noticed that there were some issues with the `CylindricalMesh.from_domain` implementation causing it to set the radius of the mesh as if the mesh were centered on the origin, but the origin is also set, so some fixes for that are now included. I've updated the method expected test results to reflect a mesh that is translated into place and matches the size of the bounding box.

I've also added an `inscribed` attribute to these methods that will indicate whether or not the mesh should inscribe or circumscribe the bounding box of the provided domain. This isn't quite an accurate term because we'd need an elliptic mesh to properly create an inscribed mesh for non-cubic bounding boxes. I'll think on a better term.

I also now need to add tests for `inscribe=True`. This is my own fault and I accept it.

pshriwise commented:
I should also note that these updates do change the behavior of the `CylindricalMesh.from_domain` method. Previously the maximum of the x and y dimensions was used to set the mesh's outer mesh. Now, in the case that the mesh is set to enclose the bounding box, the distance from the center of the x-y face to the corner of the x-y face is used.

gonuke reviewed (changes requested):
A couple of last questions...

gonuke commented:
> I opted for `enclose_domain` over `inscribe` as it's more descriptive.

(I don't see that change yet, FYI...)

paulromano reviewed (commented):
Thanks @pshriwise. I just made one update, which is to introduce a `HasBoundingBox` protocol for the domain so that if there is some future type with a `bounding_box` attribute, we don't need to change the type annotation to account for it.

One other thought -- instead of having arguments like `name`, `mesh_id`, `phi_grid_bounds`, etc. in `from_domain` that just get passed on to the normal constructor, we could just take `**kwargs` that get passed. That way if the constructor changes, we don't need to update it in two places. What do you think of that?

paulromano commented:
@pshriwise Are you OK with me taking a stab at the `**kwargs` update as part of this PR?
